src/c_api.rs

```diff
@@ -207,16 +305,415 @@ pub unsafe extern "C" fn c2pa_sign_file(
+/// * source: pointer to a CStream
+/// # Errors
+/// Returns -1 if there were errors, otherwise returns 0
+/// The error string can be retrieved by calling c2pa_error
+/// # Safety
+/// Reads from null terminated C strings
+///
+#[no_mangle]
+pub unsafe extern "C" fn c2pa_builder_add_ingredient(
+    builder_ptr: *mut C2paBuilder,
+    ingredient_json: *const c_char,
+    format: *const c_char,
+    source: *mut CStream,
+) -> c_int {
+    let mut builder: Box<C2paBuilder> = Box::from_raw(builder_ptr);
+    let ingredient_json = from_cstr_null_check_int!(ingredient_json);
+    let format = from_cstr_null_check_int!(format);
+    let result = builder.add_ingredient(&ingredient_json, &format, &mut (*source));
+    match result {
+        Ok(_builder) => {
+            Box::into_raw(builder);
+            0 as c_int
+        }
+        Err(err) => {
+            Error::from_c2pa_error(err).set_last();
+            -1
+        }
+    }
+}
+
+/// Writes an Archive of the Builder to the destination stream
+/// # Parameters
+/// * builder_ptr: pointer to a Builder
+/// * stream: pointer to a writable CStream
+/// # Errors
+/// Returns -1 if there were errors, otherwise returns 0
+/// The error string can be retrieved by calling c2pa_error
+/// # Safety
+/// Reads from null terminated C strings
+/// # Example
+/// ```c
+/// auto result = c2pa_builder_to_archive(builder, stream);
+/// if (result < 0) {
+///   printf("Error: %s\n", c2pa_error());
+/// }
+/// ```
+#[no_mangle]
+pub unsafe extern "C" fn c2pa_builder_to_archive(
+    builder_ptr: *mut C2paBuilder,
+    stream: *mut CStream,
+) -> c_int {
+    let mut builder: Box<C2paBuilder> = Box::from_raw(builder_ptr);
+    let result = builder.to_archive(&mut (*stream));
+    match result {
+        Ok(_builder) => {
+            Box::into_raw(builder);
+            0 as c_int
+        }
+        Err(err) => {
+            Error::from_c2pa_error(err).set_last();
+            -1
+        }
+    }
+}
+
+/// Creates and writes signed manifest from the C2paBuilder to the destination stream
+/// # Parameters
+/// * builder_ptr: pointer to a Builder
+/// * format: pointer to a C string with the mime type or extension
+/// * source: pointer to a CStream
+/// * dest: pointer to a writable CStream
+/// * signer: pointer to a C2paSigner
+/// * c2pa_data_ptr: pointer to a pointer to a c_uchar (optional, can be NULL)
+/// # Errors
+/// Returns -1 if there were errors, otherwise returns the size of the c2pa data
+/// The error string can be retrieved by calling c2pa_error
+/// # Safety
+/// Reads from null terminated C strings
+/// If c2pa_data_ptr is not NULL, the returned value MUST be released by calling c2pa_release_string
+/// and it is no longer valid after that call.
+#[no_mangle]
+pub unsafe extern "C" fn c2pa_builder_sign(
+    builder_ptr: *mut C2paBuilder,
+    format: *const c_char,
+    source: *mut CStream,
+    dest: *mut CStream,
+    signer: *mut C2paSigner,
+    c2pa_data_ptr: *mut *const c_uchar,
```

Comment:
How come it is a uchar as well? Isn't it an arbitrary binary blob?

Reply:
In Rust we treat it as a u8 array, so this seemed similar, but we could make it a void*.

Reply:
we need to settle on a consistent name for the binary form of a manifest store. Ingredients have/had a field called c2pa_manifest that references the manifest store. We tend to save it as a file with a .c2pa extension. I'm a bit unsettled by calling it a c2pa_manifest, since it is really a whole manifest store. I think that's where c2pa_data came from. Maybe c2pa_binary?